sdk/tables/azure-data-tables/azure/data/tables/_encoder.py

```diff
@@ -0,0 +1,112 @@
+import abc, enum
+from typing import Any, Optional, Tuple, Union, Mapping, TypeVar, Generic, Dict
+from uuid import UUID
+from datetime import datetime
+from math import isnan
+
+from ._serialize import EdmType
+from ._entity import TableEntity
+from ._deserialize import _convert_to_entity
+from ._common_conversion import _encode_base64, _to_utc_datetime
+
+_ODATA_SUFFIX = "@odata.type"
+T = TypeVar("T")
+
+
+class TableEntityEncoderABC(abc.ABC, Generic[T]):
+    def prepare_key(self, key: str) -> str:
+        """Duplicate the single quote char to escape."""
+        try:
+            return key.replace("'", "''")
+        except (AttributeError, TypeError):
+            raise TypeError('PartitionKey or RowKey must be of type string.')
+
+    def prepare_value(self, name: Optional[str], value: Any) -> Tuple[Optional[EdmType], Optional[Union[str, int, float, bool]]]:
+        """This is a utility to allow customers to reproduce our internal encoding for types like datetime, bytes, float and int64.
+        """
+        if isinstance(value, bool):
+            return None, value
+        if isinstance(value, enum.Enum):
+            return None, value.value
+        if isinstance(value, str):
+            return None, value
+        if isinstance(value, int):
+            return None, value  # TODO: Test what happens if the supplied value exceeds int32.
+        if isinstance(value, float):
+            if isnan(value):
+                return EdmType.DOUBLE, "NaN"
+            if value == float("inf"):
+                return EdmType.DOUBLE, "Infinity"
+            if value == float("-inf"):
+                return EdmType.DOUBLE, "-Infinity"
+            return EdmType.DOUBLE, value
+        if isinstance(value, UUID):
+            return EdmType.GUID, str(value)
+        if isinstance(value, bytes):
+            return EdmType.BINARY, _encode_base64(value)
+        if isinstance(value, datetime):
+            try:
+                if value.tables_service_value:
+                    return EdmType.DATETIME, value.tables_service_value
+            except AttributeError:
+                pass
+            return EdmType.DATETIME, _to_utc_datetime(value)
+        if value is None:
+            return None, None
+        if name:
+            raise TypeError(f"Unsupported data type '{type(value)}' for entity property '{name}'.")
+        raise TypeError(f"Unsupported data type '{type(value)}'.")
+
+    def encode_property(self, name: Optional[str], value: Any) -> Tuple[Optional[EdmType], Optional[Union[str, int, float, bool]]]:
```

Comment:
I'm not sure that we need this method?
It just seems to be a wrapper around `encode_value` with the extra tuple handling...

Reply:
Have moved tuple handling into `prepare_value`.